cpp/include/cuvs/neighbors/cagra.hpp

```diff
@@ -385,35 +387,65 @@ struct index : cuvs::neighbors::index {
-           const cuvs::neighbors::cagra::index_params& params,
-           raft::device_matrix_view<const uint8_t, int64_t, raft::row_major> dataset)
-  -> cuvs::neighbors::cagra::index<uint8_t, uint32_t>;
-
-auto build(raft::resources const& handle,
-           const cuvs::neighbors::cagra::index_params& params,
-           raft::host_matrix_view<const uint8_t, int64_t, raft::row_major> dataset)
-  -> cuvs::neighbors::cagra::index<uint8_t, uint32_t>;
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::device_matrix_view<const float, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
+  std::optional<float> refine_rate                                           = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::index_params> pq_build_params       = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params        = std::nullopt,
+  bool construct_index_with_dataset = true) -> cuvs::neighbors::cagra::index<float, uint32_t>;
+
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::host_matrix_view<const float, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
+  std::optional<float> refine_rate                                           = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::index_params> pq_build_params       = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params        = std::nullopt,
+  bool construct_index_with_dataset = true) -> cuvs::neighbors::cagra::index<float, uint32_t>;
+
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::device_matrix_view<const int8_t, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
+  std::optional<float> refine_rate                                           = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::index_params> pq_build_params       = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params        = std::nullopt,
+  bool construct_index_with_dataset = true) -> cuvs::neighbors::cagra::index<int8_t, uint32_t>;
+
+auto build(
+  raft::resources const& handle,
+  const cuvs::neighbors::cagra::index_params& params,
+  raft::host_matrix_view<const int8_t, int64_t, raft::row_major> dataset,
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt,
+  std::optional<float> refine_rate                                           = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::index_params> pq_build_params       = std::nullopt,
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params        = std::nullopt,
```

Comment:
These options are useful for fine tuning index construction. One thing we should consider: we could make these member of `cagra::index_params`, similarly to the `compression `struct which is part of `index_params`.